example-plugins/inj-tyfam-plugin/src/Plugin/InjTyFam.hs

```diff
@@ -0,0 +1,243 @@
+      Type.EqPred eqRel lty rty -> case eqRel of
+          Type.ReprEq -> skip   -- TODO anything useful to do here?
+          Type.NomEq  -> case splitter fskEnv lty rty of
+              Nothing         -> skip
+              Just (tc, args) -> do
+                  mbChanges <- go [] [] args
+                  case mbChanges of
+                    Nothing -> pure $ TcRnTypes.TcPluginContradiction [ct]
+                    Just x  -> ok lty tc x
+  where
+    predTree :: Type.PredTree
+    predTree = Type.classifyPredType (TcRnTypes.ctPred ct)
+
+    -- 'Nothing' means contradiction
+    -- 'Just' means (updated rtys, new constraints)
+    go :: [Xi] -> [Ct] -> [Triple] -> M (Maybe ([Xi], [Ct]))
+    go acc1 acc2 = \case
+        MkTriple lty rty i : args
+
+            -- nothing to learn at this argument position
+            | not i || Type.eqType lty rty ->
+            go (rty : acc1) acc2 args
+
+            -- contradiction!
+            | Unify.typesCantMatch [(lty, rty)] ->
+            pure Nothing
+
+            -- emit new constraint and update the RHS argument
+            | otherwise -> do
+            new <- impliedGivenEq ct lty rty
+            go (lty : acc1) (new : acc2) args
+
+        [] -> pure $ Just (reverse acc1, acc2)
+
+    -- replace the equality with one with updated RHS arguments
+    ok :: TcType -> TyCon -> ([Xi], [Ct]) -> M R
+    ok lty tc (rtys', news)
+        | null news = skip
+        | otherwise = do
+        new <- impliedGivenEq ct lty rty'
+        pure $ TcRnTypes.TcPluginOk [(ev, ct)] (new:news)
+      where
+        rty' :: TcType
+        rty' = Type.mkTyConApp tc rtys'
+
+        ev :: EvTerm
+        ev = TcRnTypes.ctEvTerm $ TcRnTypes.cc_ev ct
+
+-- | Emit a new Given equality constraint implied by another Given
+-- equality constraint
+
+impliedGivenEq :: Ct -> TcType -> TcType -> M Ct
+impliedGivenEq ct lty rty = do
+    let
+      new_co :: Coercion
+      new_co = Coercion.mkUnivCo
+        (TyCoRep.PluginProv "inj-tyfam-plugin")
+        TyCon.Nominal
+        lty
+        rty
+
+    -- TODO how to incorporate @ctEvId ct@? (see #15248 on GitLab ghc)
+    new_ev <- TcPluginM.newGiven
+      (TcRnTypes.bumpCtLocDepth (TcRnTypes.ctLoc ct))  -- TODO don't bump?
+      (Type.mkPrimEqPredRole TyCon.Nominal lty rty)
+      (CoreSyn.Coercion new_co)
+
+    pure $ TcRnTypes.mkNonCanonical new_ev
+
+-----
+
+-- | See 'mkFskEnv'
+
+type FskEnv = VarEnv (TyCon, [Xi])
+
+-- | An incremental map from the @FunEq@s
+--
+-- NOTE not necessarily idempotent
+
+mkFskEnv :: [Ct] -> FskEnv
+mkFskEnv cts =
+    VarEnv.mkVarEnv
+    [ (cc_fsk, (cc_fun, cc_tyargs))
+    | TcRnTypes.CFunEqCan{..} <- cts
+    ]
+
+unfsk :: FskEnv -> TcType -> TcType
+unfsk fskEnv t = fromMaybe t $ do
+    v <- Type.getTyVar_maybe t
+    uncurry Type.mkTyConApp <$> VarEnv.lookupVarEnv fskEnv v
+
+-----
+
+-- | A pair of arguments involved in an actionable equality constraint
+
+data Triple = MkTriple Xi Xi Bool
```

Comment:
Maybe give this a less generic name?

```haskell
data TyFamArgPair
  = MkTyFamArgPair
  { isInjArg :: !Bool
  , lhsArg :: !Xi
  , rhsArg :: !Xi
  }
```

Then you can also use puns/wildcards in the `go` function.

Reply:
Done, thanks.